File: pc1/broker_multihilo.py
# broker_multihilo.py - broker zeromq para el pc1 (ingesta de datos)
# 
# este script actúa como intermediario entre los sensores y el 
# servicio de analítica. 
# usa xsub/xpub para reenviar los mensajes sin procesarlos.
# 
# tiene un diseño multihilo:
#   - un hilo para el proxy (reenviar mensajes)
#   - un hilo para contar mensajes (métricas de rendimiento)
# 
# autores: miguel angel acuña, juan david acuña, y samuel felipe manrique - sistemas distribuidos 2026-10

import zmq
import threading
import time
import sys

# ============================================================
# CONFIGURACIÓN DE RED - Cambiar según la IP del PC1
# ============================================================
BROKER_IP = "10.43.98.198"
PUERTO_XSUB = 5555   # Aquí se conectan los sensores (PUB)
PUERTO_XPUB = 5556   # Aquí se conecta la analítica (SUB)

# TODO: Implementar CurveZMQ para la entrega final

# Variable global para el contador
contador_mensajes = 0
lock_contador = threading.Lock()

def hilo_metricas():
    # este hilo simplemente imprime el valor del contador 
    # global cada 30 segundos.
    global contador_mensajes
    print("[METRICAS] Hilo de métricas iniciado")

    while True:
        time.sleep(30)
        
        lock_contador.acquire()
        total = contador_mensajes
        contador_mensajes = 0  # Reseteo el contador
        lock_contador.release()
        
        tasa = total / 30.0
        print(f"[METRICAS] Mensajes: {total} | Tasa: {tasa:.2f} msg/s")


def hilo_proxy(contexto):
    # este hilo es el proxy principal del broker.
    # recibe mensajes de los sensores y los reenvía manualmente a 
    # la analítica.
    # no usamos zmq.proxy() para poder contar los mensajes explícitamente.
    global contador_mensajes

    # Socket XSUB - aquí llegan los mensajes de los sensores
    frontend = contexto.socket(zmq.XSUB)
    frontend.bind(f"tcp://{BROKER_IP}:{PUERTO_XSUB}")
    print(f"[BROKER] XSUB esperando sensores en tcp://{BROKER_IP}:{PUERTO_XSUB}")

    # Socket XPUB - desde aquí salen los mensajes hacia la analítica
    backend = contexto.socket(zmq.XPUB)
    backend.bind(f"tcp://{BROKER_IP}:{PUERTO_XPUB}")
    print(f"[BROKER] XPUB esperando suscriptores en tcp://{BROKER_IP}:{PUERTO_XPUB}")

    print("[BROKER] Proxy iniciado manual, reenviando mensajes...")

    while True:
        try:
            # Recibimos el mensaje crudo (string con tópico y JSON)
            mensaje = frontend.recv_string()
            
            # Sumamos 1 a la variable global del contador
            lock_contador.acquire()
            contador_mensajes += 1
            lock_contador.release()

            # Reenviamos el mensaje
            backend.send_string(mensaje)
            
        except zmq.ContextTerminated:
            print("[BROKER] Proxy detenido")
            break
        except Exception as e:
            print(f"[BROKER] Error reenviando: {e}")

    frontend.close()
    backend.close()


def main():
    print("=" * 60)
    print("  BROKER MULTIHILO - PC1 (Ingesta de datos)")
    print("=" * 60)
    print(f"  Sensores (XSUB) -> tcp://{BROKER_IP}:{PUERTO_XSUB}")
    print(f"  Analítica (XPUB) -> tcp://{BROKER_IP}:{PUERTO_XPUB}")
    print("=" * 60)

    # Creo el contexto de ZeroMQ (compartido entre hilos)
    contexto = zmq.Context()

    # Inicio el hilo de métricas (daemon para que muera con el programa)
    t1 = threading.Thread(target=hilo_metricas, daemon=True)
    t1.start()

    # Inicio el hilo del proxy
    t2 = threading.Thread(target=hilo_proxy, args=(contexto,), daemon=True)
    t2.start()

    print("[BROKER] Broker corriendo. Presione Ctrl+C para detener.\n")

    # Mantengo el programa vivo hasta que el usuario lo detenga
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("\n[BROKER] Cerrando broker...")
        contexto.term()
        print("[BROKER] Listo, broker cerrado.")


if __name__ == "__main__":
    main()

File: java/pc1/src/main/java/BrokerMultihilo.java
// broker_multihilo.java - broker zeromq para el pc1 (ingesta de datos)
//
// este script actua como intermediario entre los sensores y el
// servicio de analitica.
// usa xsub/xpub para reenviar los mensajes sin procesarlos.
//
// tiene un diseno multihilo:
//   - un hilo para el proxy (reenviar mensajes)
//   - un hilo para contar mensajes (metricas de rendimiento)
//
// autores: miguel angel acuna, juan david acuna, y samuel felipe manrique - sistemas distribuidos 2026-10

import org.zeromq.SocketType;
import org.zeromq.ZContext;
import org.zeromq.ZMQ;

import java.util.concurrent.locks.ReentrantLock;

public class BrokerMultihilo {

    // ============================================================
    // configuracion de red - cambiar segun la ip del pc1
    // ============================================================
    static String BROKER_IP = "10.43.98.198";
    static int PUERTO_XSUB = 5555;   // aqui se conectan los sensores (pub)
    static int PUERTO_XPUB = 5556;   // aqui se conecta la analitica (sub)

    // variable global para el contador
    static int contadorMensajes = 0;
    static ReentrantLock lockContador = new ReentrantLock();

    // este hilo simplemente imprime el valor del contador
    // global cada 30 segundos.
    static void hiloMetricas() {
        System.out.println("[METRICAS] Hilo de metricas iniciado");

        while (true) {
            try {
                Thread.sleep(30000);
            } catch (InterruptedException e) {
                break;
            }

            lockContador.lock();
            int total = contadorMensajes;
            contadorMensajes = 0;  // reseteo el contador
            lockContador.unlock();

            double tasa = total / 30.0;
            System.out.printf("[METRICAS] Mensajes: %d | Tasa: %.2f msg/s%n", total, tasa);
        }
    }

    // este hilo es el proxy principal del broker.
    // recibe mensajes de los sensores y los reenvia manualmente a
    // la analitica.
    // no usamos zmq.proxy() para poder contar los mensajes explicitamente.
    static void hiloProxy(ZContext contexto) {

        // socket xsub - aqui llegan los mensajes de los sensores
        ZMQ.Socket frontend = contexto.createSocket(SocketType.XSUB);
        frontend.bind("tcp://" + BROKER_IP + ":" + PUERTO_XSUB);
        System.out.println("[BROKER] XSUB esperando sensores en tcp://" + BROKER_IP + ":" + PUERTO_XSUB);

        // socket xpub - desde aqui salen los mensajes hacia la analitica
        ZMQ.Socket backend = contexto.createSocket(SocketType.XPUB);
        backend.bind("tcp://" + BROKER_IP + ":" + PUERTO_XPUB);
        System.out.println("[BROKER] XPUB esperando suscriptores en tcp://" + BROKER_IP + ":" + PUERTO_XPUB);

        System.out.println("[BROKER] Proxy iniciado manual, reenviando mensajes...");

        // uso un poller para manejar ambos sockets
        ZMQ.Poller poller = contexto.createPoller(2);
        poller.register(frontend, ZMQ.Poller.POLLIN);
        poller.register(backend, ZMQ.Poller.POLLIN);

        while (!Thread.currentThread().isInterrupted()) {
            try {
                poller.poll(1000);

                // mensajes del frontend (sensores -> broker)
                if (poller.pollin(0)) {
                    byte[] mensaje = frontend.recv(0);
                    if (mensaje != null) {
                        // sumo 1 a la variable global del contador
                        lockContador.lock();
                        contadorMensajes++;
                        lockContador.unlock();

                        // reenvio el mensaje
                        backend.send(mensaje, 0);
                    }
                }

                // suscripciones del backend (analitica -> sensores)
                if (poller.pollin(1)) {
                    byte[] suscripcion = backend.recv(0);
                    if (suscripcion != null) {
                        frontend.send(suscripcion, 0);
                    }
                }
            } catch (Exception e) {
                System.out.println("[BROKER] Proxy detenido");
                break;
            }
        }

        frontend.close();
        backend.close();
    }

    public static void main(String[] args) {
        System.out.println("============================================================");
        System.out.println("  BROKER MULTIHILO - PC1 (Ingesta de datos)");
        System.out.println("============================================================");
        System.out.println("  Sensores (XSUB) -> tcp://" + BROKER_IP + ":" + PUERTO_XSUB);
        System.out.println("  Analitica (XPUB) -> tcp://" + BROKER_IP + ":" + PUERTO_XPUB);
        System.out.println("============================================================");

        // creo el contexto de zeromq (compartido entre hilos)
        ZContext contexto = new ZContext();

        // inicio el hilo de metricas (daemon para que muera con el programa)
        Thread t1 = new Thread(BrokerMultihilo::hiloMetricas);
        t1.setDaemon(true);
        t1.start();

        // inicio el hilo del proxy
        Thread t2 = new Thread(() -> hiloProxy(contexto));
        t2.setDaemon(true);
        t2.start();

        System.out.println("[BROKER] Broker corriendo. Presione Ctrl+C para detener.\n");

        // mantengo el programa vivo hasta que el usuario lo detenga
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            System.out.println("\n[BROKER] Cerrando broker...");
            contexto.close();
            System.out.println("[BROKER] Listo, broker cerrado.");
        }));

        try {
            while (true) {
                Thread.sleep(1000);
            }
        } catch (InterruptedException e) {
            // nada
        }
    }
}
